server/api.js

```diff
@@ -77,10 +77,59 @@ async function getSensor(req, res) {
   }
 }
 
+const validateMessageClients = Validator.body(['braveKey', 'data']).exists()
+
+async function messageClients(req, res) {
```

Comment:
I can receive all locations as well as all clients with two DB calls at the beginning of the function call, then write logic to find the locations that match certain clients in JS. That would be faster and lighter on the DB; so I can get on that! 😊

Reply:
That would make a big difference. But there's something you could do that would be even faster. It has its pros and cons. What do you think it is?

Reply:
Simply write a new database query in `db/db.js`? That way it would be just one query that covers all bases, and only returns what we need (e.g., responder phone numbers).

Reply:
Yup! You got it.

And now that you have a few different options for how to implement this, what are the pros and cons of each?

Reply:
Hahaha this is fun.

**One DB call for clients, n DB calls for client's locations**
Pros:
- Simplest logic.
- The JS can be identical between Sensor / Buttons
Cons:
- Very slow, uses a lot of DB calls where unnecessary.

**Two DB calls at the beginning:**
Pros:
- No need to write an SQL query; uses existing functions.
- The JS can be identical between Sensor / Buttons
Cons:
- There is a lot of unused information from the queries; it is only the responder phone numbers that we care about, from clients that fit certain criteria.

**Write a new DB query; e.g.,** `db.getActiveClients()`:
Pros:
- Fastest; most of the logic is done by the database, which already implements fast sorting and search algorithms from a fundamental level.
- Least logic implemented in jS
Cons:
- The Sensor and Buttons DBs are _slightly_ different, so the queries will need to be a little different. (Not a big problem, though!)

Reply:
Last choice is definitely the best. I'll start working on that implementation!

Reply:
Yay! Glad you're enjoying it.

I'll add another con to the writing a new DB query. It's generally more difficult to test DB queries or stored procedures than it is to test the code that calls it.

And I'll correct the thought that the JS code would be identical. There is a `buttons` table in Buttons instead of the Sensors' `locations`. So there would still be a small modification, similar to how the SQL would need to be modified in the new DB query option.

So given all that and the context of how often this function is going to run. What option do you choose?

Reply:
Definitely still the new DB query! I feel that it won't require much testing beyond ensuring that the logic is sound, and I'll run it by you / ask any questions as I work on it, so that shouldn't be much of an issue. Additionally, this query could end up being useful later, as it is quite general.

Reply:
Sounds like a solid plan and good reasoning behind it. Thanks!